url/url.cpp

```diff
@@ -51,4 +101,1391 @@ std::string blob_url_create(Origin const &origin) {
+                state_host();
+                break;
+            case ParserState::Port:
+                state_port();
+                break;
+            case ParserState::File:
+                state_file();
+                break;
+            case ParserState::FileSlash:
+                state_file_slash();
+                break;
+            case ParserState::FileHost:
+                state_file_host();
+                break;
+            case ParserState::PathStart:
+                state_path_start();
+                break;
+            case ParserState::Path:
+                state_path();
+                break;
+            case ParserState::OpaquePath:
+                state_opaque_path();
+                break;
+            case ParserState::Query:
+                state_query();
+                break;
+            case ParserState::Fragment:
+                state_fragment();
+                break;
+            case ParserState::Failure:
+                return std::nullopt;
+            case ParserState::Terminate:
+                // I use this state where the spec returns "nothing" (i.e, the parser is modifying a given optional URL)
+                // Instead of modifying it in-place, I modify a copy and return that instead of nothing.
+                return url_;
+            default:
+                return std::nullopt;
+        }
+
+        if (is_eof()) {
+            break;
+        }
+
+        advance(1);
+    }
+
+    return url_;
+}
+
+// https://url.spec.whatwg.org/#scheme-start-state
+void UrlParser::state_scheme_start() {
+    if (util::is_alpha(peek())) {
+        buffer_ += util::lowercased(peek());
+
+        state_ = ParserState::Scheme;
+    } else if (!state_override_.has_value()) {
+        state_ = ParserState::NoScheme;
+
+        // This can underflow pos_; that's ok, because it's incremented again before it's ever used.
+        back(1);
+    } else {
+        state_ = ParserState::Failure;
+
+        return;
+    }
+}
+
+// https://url.spec.whatwg.org/#scheme-state
+void UrlParser::state_scheme() {
+    if (util::is_alphanumeric(peek()) || peek() == '+' || peek() == '-' || peek() == '.') {
+        buffer_ += util::lowercased(peek());
+    } else if (peek() == ':') {
+        if (state_override_.has_value()) {
+            if (special_schemes.contains(url_.scheme) && !special_schemes.contains(buffer_)) {
+                state_ = ParserState::Terminate;
+
+                return;
+            }
+            if (!special_schemes.contains(url_.scheme) && special_schemes.contains(buffer_)) {
+                state_ = ParserState::Terminate;
+
+                return;
+            }
+            if ((!url_.user.empty() || !url_.passwd.empty() || url_.port.has_value()) && buffer_ == "file") {
+                state_ = ParserState::Terminate;
+
+                return;
+            }
+            if (url_.scheme == "file" && url_.host.has_value() && url_.host->type == HostType::Empty) {
```

Comment:
The spec does make a distinction between a null host and an empty host. See https://url.spec.whatwg.org/#concept-url and https://url.spec.whatwg.org/#host-representation . This is probably because the Host field in the URL is optional.